hal/mbed_mpu_api.c

```diff
@@ -0,0 +1,120 @@
+#include "cmsis.h"
+
+void mbed_mpu_init()
+{
+    mbed_mpu_enable_ram_xn(false);
+}
+
+void mbed_mpu_free()
+{
+    mbed_mpu_enable_ram_xn(false);
+}
+
+void mbed_mpu_enable_ram_xn(bool enable)
+{
+#if defined (__MPU_PRESENT) && (__MPU_PRESENT == 1U)
+    const uint32_t regions = (MPU->TYPE & MPU_TYPE_DREGION_Msk) >> MPU_TYPE_DREGION_Pos;
+
+    // Disable the MCU
+    MPU->CTRL = 0;
+
+
+    // Reset all mapping
+    for (uint32_t i = 0; i < regions; i++) {
+        // Select region
+        MPU->RNR = i;
+        // Disable region
+        MPU->RASR = 0;
+        // Set address to 0
+        MPU->RBAR = 0;
+    }
+
+    if (!enable) {
+        return;
+    }
+
+    /*
+     * ARMv6m and ARMv7m memory map:
+     *
+     * Start        End            Name            Executable by default        Mbed MPU protection
+     * 0x00000000 - 0x1FFFFFFF     Code            Yes                          Execute disabled for second half
+     * 0x20000000 - 0x3FFFFFFF     SRAM            Yes                          Execute disabled
+     * 0x40000000 - 0x5FFFFFFF     Peripheral      No
+     * 0x60000000 - 0x7FFFFFFF     RAM             Yes                          Execute disabled
+     * 0x80000000 - 0x9FFFFFFF     RAM             Yes                          Execute disabled
+     * 0xA0000000 - 0xBFFFFFFF     Device          No
+     * 0xC0000000 - 0xDFFFFFFF     Device          No
+     * 0xE0000000 - 0xFFFFFFFF     System          No
+     */
+
+    if (regions >= 2) {
+        // Select region 0
+        MPU->RNR = 0;
+        // Set address to 0
+        MPU->RBAR = 0;
+        // Configure and enable region
+        MPU->RASR =
+                (1 << MPU_RASR_XN_Pos) |            // Execute Never
+                (3 << MPU_RASR_AP_Pos) |            // Full R/W for privileged and unprivileged
+                (1 << MPU_RASR_TEX_Pos) |           // Normal memory
+                (0 << MPU_RASR_S_Pos) |             // Non-shareable
+                (0 << MPU_RASR_C_Pos) |             // Normal memory
+                (0 << MPU_RASR_B_Pos) |             // Normal memory
+                (1 << (MPU_RASR_SRD_Pos + 0)) |     // Disable Sub-region
+                (0 << (MPU_RASR_SRD_Pos + 0)) |     // Enable Sub-region SRAM 0x20000000 - 0x3FFFFFFF
+                (1 << (MPU_RASR_SRD_Pos + 2)) |     // Disable Sub-region
+                (0 << (MPU_RASR_SRD_Pos + 0)) |     // Enable Sub-region RAM 0x60000000 - 0x7FFFFFFF
+                (0 << (MPU_RASR_SRD_Pos + 0)) |     // Enable Sub-region RAM 0x80000000 - 0x9FFFFFFF
+                (1 << (MPU_RASR_SRD_Pos + 5)) |     // Disable Sub-region
+                (1 << (MPU_RASR_SRD_Pos + 6)) |     // Disable Sub-region
+                (1 << (MPU_RASR_SRD_Pos + 7)) |     // Disable Sub-region
+                ((32 - 1) << MPU_RASR_SIZE_Pos) |   // 2**(32) == 4GB region size
+                (1 << MPU_RASR_ENABLE_Pos);         // Enable
+
+        // Select region 1
+        MPU->RNR = 1;
```

Comment:
It appears the Region 0 and Region 1 overlap now as they have the same base address=0. With ARMv8M architecture overlapping regions are not allowed. If we can figure out a configuration without overlapping regions this can easily be ported for ARMv8M platforms as well.  Sub-regions are also taken out from ARMv8M MPU.

Reply:
The ARMv8M MPU has a different register interface and functionality so it will need to have a separate driver regardless of this. Removing the overlap requires using an extra MPU region, so my preference would be to leave this as it is.

Reply:
This is going beyond the current scope, but I would quite like to see the ROM 00000000 region marked `ARM_MPU_AP_RO` while protection is on - would aid debugging by generating clean synchronous memory manager aborts from null-pointer-derived writes, rather than asynchronous bus errors.